Move versioning from html generator to workflow generator

cmbiwer opened the pull request:
This PR moves the call to the `versioning` module from `pycbc_make_html_page` to the workflow generator. It has been implemented in `pycbc_make_hdf_workflow`.

It is envoked by passing the path to the output dir and a `ConfigParser` object for the workflow (so that it can get all the executables).

This PR also puts meta-data in those html files to use the `render-template=render_text` so that it uses a render function that is for preformatted html/text files. Includes a `title` as well. I added the `render_text` function in `render.py` and I added a `file_pre.html` file template to do this.

A test of the HDF workflow generator versioning is here: https://sugar-jobs.phy.syr.edu/~cbiwer/tmp/html_version/versioning/

ahnitz commented:
Hey Chris, it's not immediately obvious why the version couldn't be determined for most of the executables. It looks like it is searching for an "Id: " string, which seems to be there.

spxiwh commented:
Hi Chris,

Thanks for doing this, I think this is a more sensible home for this, and gives you access to the correct configuration object.

However, you have changed the subprocess.check_output call to workflow.core.check_output. These two commands are not equivalent! Specifically, here we also need to check for stderr as well as stdout. The workflow version only checks for stdout. Many codes write the --version output to stderr and these codes now have unrecognized version strings. This causes the issue Alex mentions.

Cheers
Ian

ahnitz commented:
Ian, the issue is that subprocess.check_output doesn't exist on python 2.6

On Fri, Jun 26, 2015 at 4:13 AM, spxiwh <notifications@github.com> wrote:

> Hi Chris,
>
> Thanks for doing this, I think this is a more sensible home for this, and
> gives you access to the correct configuration object.
>
> However, you have changed the subprocess.check_output call to
> workflow.core.check_output. These two commands are not equivalent!
> Specifically, here we also need to check for stderr as well as stdout. The
> workflow version only checks for stdout. Many codes write the --version
> output to stderr and these codes now have unrecognized version strings.
> This causes the issue Alex mentions.
>
> Cheers
> Ian
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ligo-cbc/pycbc/pull/105#issuecomment-115569429>.
>

spxiwh commented:
Okay. I'll see if I can edit the workflow version to also optionally return stderr for this case.

spxiwh commented:
Please see the pull request above to add some patches to Chris' remote branch. I think this fixes the problem and differentiates between "code doesn't have --version" and "code has --version but I don't understand what it means".

spxiwh commented:
It looks like there's now some conflict between this and origin/master. This probably needs to be rebased, conflicts resolved and then force-pushed back to this branch.

cmbiwer commented:
>  "code doesn't have --version" and "code has --version but I don't understand what it means"
I think the problem still exists.

`check_output_error_and_retcode` does not have the new line `\n` that the `versioning.py` code expects. So the `.split('\n')` in `versioning.py` at line 174 does nothing useful.

In ipython:
```
check_output_error_and_retcode(['/home/cbiwer/pycbc/version_dev_20150625/bin/pycbc_inspiral', '--version'])
('',
 'Branch: version_dev Tag: None Id: 965b38dacb74ecd30f555253f673fb8b510aa958 Builder: Christopher M. Biwer <christopher.biwer@ligo.org> Build date: 2015-06-26 12:23:00\n+0000 Repository status is UNCLEAN: Modified working tree\n',
 0)
```

cmbiwer commented:
I fixed the conflicts. And it works now, see:

https://sugar-jobs.phy.syr.edu/~cbiwer/tmp/html_version/versioning/

cmbiwer commented:
Travis CI failure is actually:
```
Your test run exceeded 50 minutes. 

One possible solution is to split up your test run.
```

spxiwh commented:
I'm happy for you to merge this when ready.
